src/Data/Vec/Properties.agda

```diff
@@ -691,8 +696,209 @@ foldr-fusion : ∀ {A : Set a}
+  ∷ʳ-injective []          []          refl = (refl , refl)
+  ∷ʳ-injective (x ∷ xs)    (y  ∷ ys)   eq   with ∷-injective eq
+  ... | refl , eq′ = Prod.map (P.cong (x ∷_)) id (∷ʳ-injective xs ys eq′)
+
+  ∷ʳ-injectiveˡ : ∀ {n} (xs ys : Vec A n) → xs ∷ʳ x ≡ ys ∷ʳ y → xs ≡ ys
+  ∷ʳ-injectiveˡ xs ys eq = proj₁ (∷ʳ-injective xs ys eq)
+
+  ∷ʳ-injectiveʳ : ∀ {n} (xs ys : Vec A n) → xs ∷ʳ x ≡ ys ∷ʳ y → x ≡ y
+  ∷ʳ-injectiveʳ xs ys eq = proj₂ (∷ʳ-injective xs ys eq)
+
+
+------------------------------------------------------------------------
+-- _∷ʳ_ and map
+
+∷ʳ-map-commute : (f : A → B) → ∀ {n} x (xs : Vec A n) →
+                 map f (xs ∷ʳ x) ≡ (map f xs) ∷ʳ (f x)
+∷ʳ-map-commute f x [] = refl
+∷ʳ-map-commute f x (y ∷ xs) = P.cong (f y ∷_) (∷ʳ-map-commute f x xs)
+
+------------------------------------------------------------------------
+-- _∷ʳ_, _ʳ++_ and reverse
+
+-- reverse of cons is snoc of reverse.
+
+unfold-reverse : ∀ {n} (x : A) xs → reverse (x ∷ xs) ≡ reverse {n = n} xs ∷ʳ x
+unfold-reverse x xs =  P.sym (foldl-fusion (_∷ʳ x) [] [ x ] refl (λ b x → refl) xs)
+
+unfold-ʳ++ : ∀ {m n} {xs : Vec A m} {ys : Vec A n} → xs ʳ++ ys ≡ reverse xs ++ ys
+unfold-ʳ++ {A = A} {m} {n} {xs} {ys} =
+  P.sym (foldl-fusion (_++ ys) [] ys refl (λ b x → refl) xs)
+
+-- reverse and map
+
+reverse-map-commute : (f : A → B) →
+                      ∀ {n} (xs : Vec A n) → map f (reverse xs) ≡ reverse (map f xs)
+reverse-map-commute f [] = refl
+reverse-map-commute f (x ∷ xs) = begin
+  map f (reverse (x ∷ xs))
+    ≡⟨ P.cong (map f) (unfold-reverse x xs) ⟩
+  map f (reverse xs ∷ʳ x)
+    ≡⟨ ∷ʳ-map-commute f x (reverse xs) ⟩
+  map f (reverse xs) ∷ʳ f x
+    ≡⟨ P.cong (_∷ʳ f x) (reverse-map-commute f xs ) ⟩
+  reverse (map f xs) ∷ʳ f x
+    ≡⟨ P.sym (unfold-reverse (f x) (map f xs)) ⟩
+  reverse (f x ∷ map f xs)
+    ≡⟨⟩
+  reverse (map f (x ∷ xs)) ∎
+    where open P.≡-Reasoning
+
+-- _ʳ++_ and map
+-- map distributes over reverse-append.
+
+map-ʳ++ : ∀ {A : Set a} {B : Set b} {m n} (f : A → B) (xs : Vec A m) {ys : Vec A n} →
+          map f (xs ʳ++ ys) ≡ map f xs ʳ++ map f ys
+map-ʳ++ f xs {ys} = begin
+  map f (xs ʳ++ ys)
+    ≡⟨ P.cong (map f) (unfold-ʳ++ {xs = xs} {ys = ys}) ⟩
+  map f (reverse xs ++ ys)
+    ≡⟨ map-++-commute f (reverse xs) ys ⟩
+  map f (reverse xs) ++ map f ys
+    ≡⟨ P.cong (_++ map f ys) (reverse-map-commute f xs) ⟩
+  reverse (map f xs) ++ map f ys
+    ≡⟨ P.sym (unfold-ʳ++ {xs = map f xs} {ys = map f ys}) ⟩
+  map f xs ʳ++ map f ys ∎
+    where open P.≡-Reasoning
+
+-- reverse and foldr
+
+reverse-foldr : ∀ {B : ℕ → Set b} {n} (f : ∀ {n} → A → B n → B (suc n)) b →
+                foldr B f b ∘ reverse ≗ foldl B {n} (λ {n} → flip (f {n})) b
+reverse-foldr {A = A} {B = B} f b xs =
+  foldl-fusion {B = Vec A} (foldr B f b) [] b refl (λ b x → refl) xs
+
+-- reverse and foldl
+
+reverse-foldl : ∀ {B : ℕ → Set b} {n} (f : ∀ {n} → B n → A → B (suc n)) b →
+                foldl B {n} f b ∘ reverse ≗ foldr B (λ {n} → flip (f {n})) b
```

Comment:
Is there a reason that the equality in this is flipped compared to `reverse-foldr`? They should probably be consistent?

Reply:
So I think that `reverse-foldr` and `reverse-foldl` are consistent now: both of these lemmas now put the length argument on the left of the equality, and each has the 'flipped' function on the right. I have made considerable reorganisation of the `Properties` file, putting properties of `foldl` *before* those of `foldr`, and making heavy use of anonymous modules to group lemmas together. But there may still be opportunities for improvement!

Reply:
And they have each had their names flipped now: `foldr-reverse`, `foldl-reverse`.

Reply:
Thanks, definitely the right way to go!